src/services/jsTyping.ts

```diff
@@ -0,0 +1,286 @@
+            for (const notFoundTypingName of notFoundTypingNames) {
+                if (inferredTypings.hasOwnProperty(notFoundTypingName) && !inferredTypings[notFoundTypingName]) {
+                    delete inferredTypings[notFoundTypingName];
+                }
+            }
+
+            // The "installed" property in the tsd.json serves as a registry of installed typings. Each item
+            // of this object has a key of the relative file path, and a value that contains the corresponding
+            // commit hash.
+            if (hasProperty(tsdJsonDict, "installed")) {
+                for (const cachedTypingPath in tsdJsonDict.installed) {
+                    // Assuming the cachedTypingPath has the format of "[package name]/[file name]"
+                    const cachedTypingName = cachedTypingPath.substr(0, cachedTypingPath.indexOf("/"));
+                    // If the inferred[cachedTypingName] is already not null, which means we found a corresponding
+                    // d.ts file that coming with the package. That one should take higher priority.
+                    if (hasProperty(inferredTypings, cachedTypingName) && !inferredTypings[cachedTypingName]) {
+                        inferredTypings[cachedTypingName] = ts.combinePaths(typingsPath, cachedTypingPath);
+                    }
+                }
+            }
+        }
+
+        // Remove typings that the user has added to the exclude list
+        for (const excludeTypingName of exclude) {
+            delete inferredTypings[excludeTypingName];
+        }
+
+        const newTypingNames: string[] = [];
+        const cachedTypingPaths: string[] = [];
+        for (const typing in inferredTypings) {
+            if (inferredTypings[typing] !== undefined) {
+                cachedTypingPaths.push(inferredTypings[typing]);
+            }
+            else {
+                newTypingNames.push(typing);
+            }
+        }
+        return { cachedTypingPaths, newTypingNames, filesToWatch };
+
+        /**
+         * Merge a given list of typingNames to the inferredTypings map
+         */
+        function mergeTypings(typingNames: string[]) {
+            if (!typingNames) {
+                return;
+            }
+
+            for (const typing of typingNames) {
+                if (!inferredTypings.hasOwnProperty(typing)) {
+                    inferredTypings[typing] = undefined;
+                }
+            }
+        }
+
+        /**
+         * Get the typing info from common package manager json files like package.json or bower.json
+         */
+        function getTypingNamesFromJson(jsonPath: string, filesToWatch: string[]) {
+            const jsonDict = tryParseJson(jsonPath, host);
+            if (jsonDict) {
+                filesToWatch.push(jsonPath);
+                if (jsonDict.hasOwnProperty("dependencies")) {
+                    mergeTypings(Object.keys(jsonDict.dependencies));
+                }
+            }

```

Comment:
No problem. Seemed like you'd want to cover those anyway, since there'll never be overlap in any of these fields and they are treated the same as regular dependencies on a local project. The only difference is when they're used as a sub-dependency dependencies.